### ISSUE TITLE
Cannot get catalog info from TinyOPDS server

### ISSUE THREAD

**Issue by slow-mo** [koreader/koreader#6088]
* KOReader version: v2020.04.1-4-gfdc92e9_2020-04-23
* Device: Kindle Paperwhite 2

#### Issue
I get a "Cannot get catalog info from" error when trying to connect to a local TinyOPDS server.
Commenting out lines 314-316 opdsbrowser.lua works around the issue. 

#### Steps to reproduce
1. Download (https://archive.codeplex.com/?p=tinyopds) and start tinyopds on a local machine
2. Add an opds catalog on your koreader device pointing to the local server
3. Open the catalog 

##### `crash.log`
```
04/24/20-11:36:28 INFO  request {
    ["url"] = "http://192.168.1.166:8080/opds",
    ["method"] = "HEAD",
    ["headers"] = {
        ["Authorization"] = "Basic Og==",
        ["Host"] = "192.168.1.166"
    },
    ["sink"] = function: 0x44eba708
}
04/24/20-11:36:30 INFO  cannot get catalog info from http://192.168.1.166:8080/opds frontend/ui/widget/opdsbrowser.lua:315: closed
```

**Comment by Frenzie** [comment]
Hold on, lines 314-316? That doesn't seem to make sense. :-/

https://github.com/koreader/koreader/blob/fdc92e950ed5a18eca7818ecdbd47852834ea07f/frontend/ui/widget/opdsbrowser.lua#L312-L323

**Comment by slow-mo** [comment]
Yep those lines. Seems to have something to do with response headers from tinyopds?
Very limited knowledge here, but it sends a Content-Type and a Connection: close header as far as I can tell.

**Comment by NiLuJe** [comment]
That vaguely looks like a "remote closed the connection" error.

What does the TineOPD logs say about that?

**Comment by slow-mo** [comment]
Unfortunately nothing at all. The HTTP GET request is logged only if those lines are commented out:
`04/24/2020 15:53:41.6	I	HttpProcessor	HTTP GET request from 192.168.1.121: /opds`

**Comment by NiLuJe** [comment]
Logging a request without its status is highly useful logging (not).

Can that be tweaked (perhabs verbose logging? debug logging?)

**Comment by slow-mo** [comment]
Nope that's the highest debug level. But maybe this can help: if I use fiddler2 as a reverse proxy, koreader can get the opds catalog from tinyopds without the "workaround". 
Fiddler2 records the following 2 sessions:
```

HEAD http://192.168.1.166/opds HTTP/1.1
User-Agent: LuaSocket 3.0-rc1
TE: trailers
Authorization: Basic Og==
Connection: close, TE
Host: 192.168.1.166


HTTP/1.1 504 Fiddler - Receive Failure
Date: Fri, 24 Apr 2020 14:45:43 GMT
Content-Type: text/html; charset=UTF-8
Connection: close
Cache-Control: no-cache, must-revalidate
Timestamp: 16:45:43.528

[Fiddler] ReadResponse() failed: The server did not return a complete response for this request. Server returned 0 bytes.                                                                                                                                                                                                                                                                                                                                                                                                       

------------------------------------------------------------------

GET http://192.168.1.166/opds HTTP/1.1
User-Agent: LuaSocket 3.0-rc1
TE: trailers
Authorization: Basic Og==
Connection: close, TE
Host: 192.168.1.166


HTTP/1.1 200 OK
Content-Type: application/atom+xml;charset=utf-8
Connection: close

<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom" xmlns:dc="http://purl.org/dc/terms/" xmlns:os="http://a9.com/-/spec/opensearch/1.1/" xmlns:opds="http://opds-spec.org/2010/catalog">
  <id>tag:root</id>
  <title>Home OPDS Server</title>
  <subtitle>running on TinyOPDS server version 2.0</subtitle>
  <updated>2020-04-24T14:45:43.7830523Z</updated>
  <icon>/favicon.ico</icon>
  <link href="/opds-opensearch.xml" rel="search" type="application/opensearchdescription+xml" />
  <link href="/opds/search?searchTerm={searchTerms}" rel="search" type="application/atom+xml" />
  <link href="/opds" rel="start" type="application/atom+xml" />
  <link href="/opds" rel="self" type="application/atom+xml" />
  <entry>
    <updated>2020-04-24T14:45:43.7986489Z</updated>
    <id>tag:root:new</id>
    <title type="text">New books (by date added)</title>
    <content type="text">1991 new books ordered by date</content>
    <link href="/opds/newdate" type="application/atom+xml;profile=opds-catalog" />
  </entry>
  <entry>
    <updated>2020-04-24T14:45:43.8298965Z</updated>
    <id>tag:root:new</id>
    <title type="text">New books (alphabetically)</title>
    <content type="text">1991 new books ordered alphabetically</content>
    <link href="/opds/newtitle" type="application/atom+xml;profile=opds-catalog" />
  </entry>
  <entry>
    <updated>2020-04-24T14:45:43.8455196Z</updated>
    <id>tag:root:authors</id>
    <title type="text">By authors</title>
    <content type="text">1991 books by 1290 authors</content>
    <link href="/opds/authorsindex" type="application/atom+xml;profile=opds-catalog" />
  </entry>
  <entry>
    <updated>2020-04-24T14:45:43.8455196Z</updated>
    <id>tag:root:sequences</id>
    <title type="text">By series</title>
    <content type="text">1991 books by 0 series</content>
    <link href="/opds/sequencesindex" type="application/atom+xml;profile=opds-catalog" />
  </entry>
  <entry>
    <updated>2020-04-24T14:45:43.8676519Z</updated>
    <id>tag:root:genre</id>
    <title type="text">By genres</title>
    <content type="text">Books grouped by genres</content>
    <link href="/opds/genres" type="application/atom+xml;profile=opds-catalog" />
  </entry>
</feed>
```

**Comment by NiLuJe** [comment]
Yup, that's quite likely the 504 we're hitting, too (i.e., TineOPDS is b0rked).

I imagine LuaSockets drops everything on 504, while fiddler attempts to retransmit whatever it got before EoT. It's quite likely the transmission is actually incomplete, though.

TL;DR: Not our bug.

**Comment by slow-mo** [comment]
That's unfortunate though, due to LuaSockets being so picky, it seems that koreader can't work with any lightweight/portable file transfer app. I had similar problems with tinyftp and the like.
I understand these programs are maybe a little bit... rough, but they somehow do work with other clients.
Not a big issue btw, calibre is slower to start but does the job.

**Comment by NiLuJe** [comment]
Okay, we're actually asking for a HEAD request.

The HEAD request appears to be to query authentication (c.f., OPDSBrowser:parseFeed).
I'm not familiar about this bit of code (or HTTP handling in general), so if someone wants to figure what's valid or not, be my guest.

I'd also double-check with curl whether it's just us/this request in particular, or if it's actually any HEAD requests that get mangled.

**Comment by slow-mo** [comment]
It's any HEAD requests that get mangled. curl -I gives:
`curl: (52) Empty reply from server`
while, with fiddler in the middle, we get the above mentioned:
```
HTTP/1.1 502 Fiddler - Connection Failed
Date: Fri, 24 Apr 2020 20:23:17 GMT
Content-Type: text/html; charset=UTF-8
Connection: close
Cache-Control: no-cache, must-revalidate
Timestamp: 22:23:17.325
```
So, without fiddler, headers are nil and the condition in opdsbrowser.lua line 314 is met. 
It's not that fiddler proceeds with a GET, it's that it fills the headers that prevent koreader erroring on line 315.

**Comment by slow-mo** [comment]
Ok, so I deleted the previous message 'cause I messed with fiddler, but it seems it was substantially correct.
It seems that any HEAD requests get mangled. curl -I gives:
`curl: (52) Empty reply from server`

Using fiddler as a reverse proxy I get the above mentioned:
```
HTTP/1.1 504 Fiddler - Receive Failure
Date: Fri, 24 Apr 2020 21:37:11 GMT
Content-Type: text/html; charset=UTF-8
Connection: close
Cache-Control: no-cache, must-revalidate
Timestamp: 23:37:11.897
```

So, with fiddler, headers are not nil and the condition in opdsbrowser.lua line 314 is not met.
As far as I can tell, the point is not that fiddler is still proceeding with a GET, but that it is filling the headers thus preventing koreader from erroring on line 315.
Comment at line 313 states that headers are checked to see if network is unavailable. As a total ignorant: is there any other convenient way to do this so that no error is raised when the server fails to reply properly to a HEAD request?

**Comment by Frenzie** [comment]
I don't think that's an accurate analysis. It's just a slightly incorrect assumption on the part of the comment writer that the only reason for this problem would be some kind of network shenanigans.

**Comment by slow-mo** [comment]
Look, I'm just an user so I'm absolutely in no position to stand by my assumptions. Maybe somebody could try and reproduce the issue at least to validate what's going on?
It's just a 9.4 MB zip file to download and unzip, no installation required, so it would really require a windowa machine and a few minutes.

**Comment by Frenzie** [comment]
Indeed, the comment that this check is to "raise error message when network is unavailable."

In my understanding of the HTTP spec as well as glancing at our code that should be read as "raise error message [in case of this problem, which most likely indicates that the] network is unavailable."

**Comment by Frenzie** [comment]
To put it another way, it's a sanity check to prevent problems, not actually a network check.

**Comment by slow-mo** [comment]
Yeah I see.

**Comment by hius07** [comment]
New reports and complaints on this.
Some local OPDS servers (TinyOPDS, freelib) do not return headers in the GET request but do return the content.
Without a header KOReader does not allow to parse the feed:
https://github.com/koreader/koreader/blob/1c42641733b59b4fdea0df8a5c4f102790dbeb52/plugins/opds.koplugin/opdsbrowser.lua#L280-L289
Some other readers (Moon+, AlReader) ignore the headers absence and work with these servers.

OPDSBrowser uses the header to generate the cache hash only, no other purposes.
https://github.com/koreader/koreader/blob/1c42641733b59b4fdea0df8a5c4f102790dbeb52/plugins/opds.koplugin/opdsbrowser.lua#L317-L321

I think we can unblock the issue and remove lines 280-282.
Network problems would be detected with the check of `code == 200` in line 286.

**Comment by NiLuJe** [comment]
I'd possibly keep the check, but just change the priority? (i.e., check code first).
